frontend/document/filemanagerdocument.lua

```diff
@@ -0,0 +1,85 @@
+local Document = require("document/document")
+local util = require("util")
+
+local FileManagerDocument = Document:new{
+    _document = false,
+    provider = "filemanagerdocument",
+    provider_name = "File Manager",
+    providers = {},
+}
+
+function FileManagerDocument:init() end
+function FileManagerDocument:close() end
+function FileManagerDocument:register() end
+
+-- opens a file with the provider with highest priority for that extension
+function FileManagerDocument:open(file)
+    local extension = util.getFileNameSuffix(file)
+    if not self.providers[extension] then return end
+    local max_priority = 0
+    local max_priority_index = 1
+    for i, v in ipairs(self.providers[extension]) do
+        if v.priority > max_priority then
+            max_priority = v.priority
+            max_priority_index = i
+        end
+    end
+    self.providers[extension][max_priority_index].open_func(file)
+end
+
+function FileManagerDocument:addHandler(name, t)
+    assert(type(name) == "string", "string expected")
+    assert(type(t) == "table", "table expected")
+    local extension, mimetype, priority, open_func
+    for k, v in pairs(t) do
+        if type(v) == "table" then
+            if type(k) == "string" then
+                extension = k
+            end
+            if type(v.mimetype) == "string" then
+                mimetype = v.mimetype
+            end
+            if type(v.open_func) == "function" then
+                open_func = v.open_func
+            end
+            priority = v.priority or 20
+        end
+
+        if extension and mimetype and open_func then
+            if not self.providers[extension] then
+                self.providers[extension] = {}
+            end
+            table.insert(self.providers[extension], v)
+            require("document/documentregistry"):addProvider(extension, mimetype, self, priority)
+        end
+    end
+end
+
+function FileManagerDocument:changeHandler(extension, t)
+    self:deleteHandler(extension)
+    self:addHandler(extension, t)
+end
+
+-- todo: remove from self.providers and documentregistry
+function FileManagerDocument:deleteHandler(extension)
+end
+
+
+function FileManagerDocument:getProps()
+    local _, _, docname = self.file:find(".*/(.*)")
+    docname = docname or self.file
+    return {
+        title = docname:match("(.*)%."),
+    }
+end
+
+function FileManagerDocument:_readMetadata()
+    Document._readMetadata(self)
+    return true
+end
+
+function FileManagerDocument:getCoverPageImage()
+    return nil
```

Comment:
Indeed, I can try to make it working now. As I said I'm not familiar enough with `Document`'s

Is it ok to return the path of the svg file here, in case there's one?

Reply:
No, it is expected to return a bb.
And `RenderImage:renderSVGImageFile(filename, width, height, zoom)` can do just that.

Reply:
I'm afraid we need to hack `bookinfomanager.lua` a little bit for it to happen.

I say I'm afraid because I'm completely ignorant of the code there.

I can say just two things ATM:

in `BookInfoManager:getBookInfo` if I return early when provider.provider is a filemanagerdocument a bookinfo table with `ignore_cover = "Y"` everything works but there's no image.

If I comment `ignore_cover`  then `getBookInfo` is called again and again.

Attaching the diff of the code I'm playing with

```diff
diff --git a/plugins/coverbrowser.koplugin/bookinfomanager.lua b/plugins/coverbrowser.koplugin/bookinfomanager.lua
index 003f59eb..902bd16d 100644
--- a/plugins/coverbrowser.koplugin/bookinfomanager.lua
+++ b/plugins/coverbrowser.koplugin/bookinfomanager.lua
@@ -317,7 +317,7 @@ end
 -- Bookinfo management
 function BookInfoManager:getBookInfo(filepath, get_cover)
     local directory, filename = util.splitFilePathName(filepath)
-
+    logger.info("getBookInfo")
     -- CoverBrowser may be used by PathChooser, which will not filter out
     -- files with unknown book extension. If not a supported extension,
     -- returns a bookinfo like-object enough for a correct display and
@@ -344,6 +344,33 @@ function BookInfoManager:getBookInfo(filepath, get_cover)
         }
     end
 
+    if DocumentRegistry:getProvider(filepath).provider == "filemanagerdocument" then
+        logger.info("FM document returns early")
+        local bookinfo = {
+            directory = directory,
+            filename = filename,
+            in_progress = 0,
+            cover_fetched = "Y",
+            has_meta = nil,
+            has_cover = true,
+            ignore_meta = "Y",
+            --ignore_cover = "Y",
+            _is_directory = false,
+            _no_provider = false,
+        }
+
+        local icon = require("document/filemanagerdocument"):getIcon(filepath)
+        logger.info("file", filepath, "got icon", icon)
+
+        if icon then
+            local cover_bb = require("ui/renderimage"):renderSVGImageFile(icon, 100, 100, 1)
+            cover_bb:setAllocated(1)
+            bookinfo["cover_bb"] = cover_bb
+        end
+
+        logger.info("cover blitbuffer", bookinfo.cover_bb)
+        return bookinfo
+    end
     self:openDbConnection()
     local row = self.get_stmt:bind(directory, filename):step()
     -- NOTE: We do not reset right now because we'll be querying a BLOB,
@@ -397,6 +424,7 @@ function BookInfoManager:getBookInfo(filepath, get_cover)
 end
 
 function BookInfoManager:extractBookInfo(filepath, cover_specs)
+    logger.info("extractBookInfo", filepath)
     -- This will be run in a subprocess
     -- We use a temporary directory for cre cache (that will not affect parent process),
     -- so we don't fill the main cache with books we're not actually reading
```

Reply:
There shouldn't be anything specific to hack in CoverBrowser.
The bit you took inspiration from is for directories.

You made your files be "FilemanagerDocument", so a document. They will go through `BookInfoManager:extractBookInfo(filepath)` where your FilemanagerDocument regular methods should be called:
```lua
DocumentRegistry:openDocument(filepath)
pages = document:getPageCount() -- remember I said this one may crash, may be the reason it did not work?
local props = document:getProps()
local cover_bb = document:getCoverPageImage()
document:close()
```
So, if you have defined those, they should be called to get (and cache) what's necessary to draw the coverbrowser items.

Reply:
thanks for the pointers.

we still need to hack bookinfomanager because the cover image cannot be derived from the filemanagerdocument without knowing which kind of extension the current `filename` has.

Reply:
Probably the job of filemanagerdocument to find out what it's main provider is ? or find out in its docsetting what was the last/default provider use (if that is saved?) or in the global settings what the default provider for that extension is.

> cannot be derived from the filemanagerdocument without knowing which kind of extension the current filename has.

Document (and its derived ones) has a self.file member.

Reply:
Thanks!

> Document (and its derived ones) has a self.file member.

That's cool!

> Probably the job of filemanagerdocument to find out what it's main provider is ? or find out in its docsetting what was the last/default provider use (if that is saved?) or in the global settings what the default provider for that extension is.

The first intention was that. ie: store priority of handlers/providers internally and open stuff with the nº 1. But I was not sure because future usages would attempt to escalate priorities to be placed on top of the rest. So finally I removed priority stuff and decided it was easier and leaner to show a popup that lets the user choose between handlers.

Not sure if it does worth the effort to add priorities again just for the icons or, instead, just iterate over them and return the blitbuffer of the first svg for each extension.